stout/collect.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <functional>
+
+#include "stout/loop.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename, typename = void>
+struct HasEmplaceBack : std::false_type {
+};
+
+template <typename T>
+struct HasEmplaceBack<
+    T,
+    std::void_t<decltype(std::declval<T>().emplace_back(
+        std::declval<typename T::value_type&&>()))>>
+  : std::true_type {
+};
+
+} // namespace detail
+
+////////////////////////////////////////////////////////////////////////
+
+template <typename Container, typename F>
+auto Collect(F f) {
+  using T = typename Container::value_type;
+
+  struct DataContainer {
+    std::function<bool(T)> func;
+    Container content;
+  };
+
+  return Loop<Container>()
+      .context(DataContainer{std::move(f), Container()})
+      .body([](auto& data, auto& stream, auto&& value) {
+        if (data.func(value)) {
```

Comment:
I'll echo what Ben said in [his comment on your previous PR](https://github.com/3rdparty/stout-eventuals/pull/29#pullrequestreview-751912487): let's not implement `Filter` functionality in `Collect`.

`Collect` should collect _all_ the values, it shouldn't ask the user whether a value should be collected or not. If the user wants only some of the values collected, they should first do a `Filter`, then a `Collect`.